lib/bolt/fiber_executor.rb

```diff
@@ -0,0 +1,141 @@
+
+    def initialize
+      @logger = Bolt::Logger.logger(self)
+      @id = 0
+      @plan_futures = []
+    end
+
+    # Whether there is more than one fiber running in parallel.
+    #
+    def in_parallel
+      plan_futures.length > 1
+    end
+
+    # Creates a new Puppet scope from the current Plan scope so that variables
+    # can be used inside the block and won't interact with the outer scope.
+    # Then creates a new Fiber to execute the block, wraps the Fiber in a
+    # Bolt::PlanFuture, and returns the Bolt::PlanFuture.
+    #
+    def create_future(scope: nil, name: nil)
+      newscope = nil
+      if scope
+        # Save existing variables to the new scope before starting the future
+        # itself so that if the plan returns before the backgrounded block
+        # starts, we still have the variables.
+        newscope = Puppet::Parser::Scope.new(scope.compiler)
+        local = Puppet::Parser::Scope::LocalScope.new
+
+        # Compress the current scopes into a single vars hash to add to the new scope
+        scope.to_hash.each_pair { |k, v| local[k] = v }
+        newscope.push_ephemerals([local])
+      end
+
+      # Create a new Fiber that will execute the provided block.
+      future = Fiber.new do
+        # Yield the new scope - this should be ignored by the block if
+        # `newscope` is nil.
+        yield newscope
+      end
+
+      # PlanFutures are assigned an ID, which is just a global incrementing
+      # integer. The main plan should always have ID 0.
+      @id += 1
+      future = Bolt::PlanFuture.new(future, @id, name)
+      @logger.trace("Created future #{future.name}")
+
+      # Register the PlanFuture with the FiberExecutor to be executed
+      plan_futures << future
+      future
+    end
+
+    # Visit each PlanFuture registered with the FiberExecutor and resume it.
+    # Fibers will yield themselves back, either if they kicked off a
+    # long-running process or if the current long-running process hasn't
+    # completed. If the Fiber finishes after being resumed, store the result in
+    # the PlanFuture and remove the PlanFuture from the FiberExecutor.
+    #
+    def round_robin
+      plan_futures.each do |future|
+        # If the Fiber is still running and can be resumed, then resume it
+        @logger.trace("Checking future '#{future.name}'")
+        if future.alive?
+          @logger.trace("Resuming future '#{future.name}'")
+          future.resume
+        end
+
+        # Once we've restarted the Fiber, check to see if it's finished again
+        # and cleanup if it has.
+        next if future.alive?
+        @logger.trace("Cleaning up future '#{future.name}'")
+
+        # If the future errored and the main plan has already exited, log the
+        # error at warn level.
+        unless plan_futures.map(&:id).include?(0) || future.state == "done"
+          Bolt::Logger.warn('errored-futures', "Error in future '#{future.name}': #{future.value}")
```

Comment:
This will log `RunFailure` error messages which state the plan was aborted. This is a bit misleading since errors in backgrounded blocks don't actually abort the plan. Here's some example output:

```
Starting: plan bolt
Starting: task facts on nix1, nix2, win
Starting: command 'tree' on nix1, nix2, win
Starting: command 'sleep 5' on nix1, nix2, win
Finished: command 'tree' with 2 failures in 1.69 sec
Error in future '2': Plan aborted: run_command 'tree' failed on 2 targets [ID: errored-futures]
skip_cleanup [ID: Skipping cleanup of tmpdir 'C:\Users\Administrator\AppData\Local\Temp\ld1ioqpz.gx1']
Finished: task facts with 0 failures in 3.64 sec
Finished: plan bolt in 4.07 sec
Finished on nix1:
  ...
Finished on nix2:
  ...
Finished on win:
  ...
Successful on 3 targets: nix1,nix2,win
Ran on 3 targets
```

Reply:
Ugh, I think this is actually going to be trickier to work around than I thought....It's hard to know when we raise the RunFailure in the plan functions if the function is being run in a background block, and other teams are [raising RunFailure errors](https://github.com/puppetlabs/pe-bolt-modules/blob/52a1c9a16aa036458524b527eb4f00177614cf24/pe-bolt-modules/pe_boltlib/lib/puppet/functions/apply_prep.rb#L23) so modifying the function signature beyond adding an optional parameter isn't a good optoin.

Reply:
I think this is an improvement we can make, but doing it right is going to be a bit complicated. I'd like to get this in first and do that as a followup to avoid scope-creep too much.

Reply:
<5 hours later>